### ISSUE TITLE
Bundler 2.2.17 fails with "Could not find sorbet-static-0.5.6403-universal-darwin-20 in any of the sources" on Linux on TruffleRuby

### ISSUE THREAD

**Issue by eregon** [rubygems/rubygems#4588]
### Describe the problem as clearly as you can

When running on truffleruby-head/dev, which you can install with ` ruby-build truffleruby-dev ~/.rubies/truffleruby-dev`.
There is a change to mark the binary-only `sorbet-static` gem as fine to install on TruffleRuby, this is needed to reproduce, it should merge soon, I'll ping here when it's done and there is a build with it.

Using this Gemfile:
```ruby
source 'https://rubygems.org'

gem 'sorbet', '0.5.6403'
gem 'libv8', '8.4.255.0'
```

The resolution seems to go wrong, because the generated Gemfile.lock on TruffleRuby is:
```
GEM
  remote: https://rubygems.org/
  specs:
    libv8 (8.4.255.0)
    sorbet (0.5.6403)
      sorbet-static (= 0.5.6403)
    sorbet-static (0.5.6403-universal-darwin-14)
    sorbet-static (0.5.6403-universal-darwin-15)
    sorbet-static (0.5.6403-universal-darwin-16)
    sorbet-static (0.5.6403-universal-darwin-17)
    sorbet-static (0.5.6403-universal-darwin-18)
    sorbet-static (0.5.6403-universal-darwin-19)
    sorbet-static (0.5.6403-universal-darwin-20)
    sorbet-static (0.5.6403-x86_64-linux)

PLATFORMS
  ruby

DEPENDENCIES
  libv8 (= 8.4.255.0)
  sorbet (= 0.5.6403)

BUNDLED WITH
   2.2.17
```

Compare that to CRuby where it is:
```
GEM
  remote: https://rubygems.org/
  specs:
    libv8 (8.4.255.0-x86_64-linux)
    sorbet (0.5.6403)
      sorbet-static (= 0.5.6403)
    sorbet-static (0.5.6403-x86_64-linux)

PLATFORMS
  x86_64-linux

DEPENDENCIES
  libv8 (= 8.4.255.0)
  sorbet (= 0.5.6403)

BUNDLED WITH
   2.2.17
```

So when doing `bundle install` on Bundler 2.2.17 on TruffleRuby it fails with:
```
$ bundle install --verbose
Running `bundle install --verbose` with bundler 2.2.17
Found no changes, using resolution from the lockfile
The definition is missing ["sorbet-static-0.5.6403-universal-darwin-20", "sorbet-static-0.5.6403-universal-darwin-19", "sorbet-static-0.5.6403-universal-darwin-18", "sorbet-static-0.5.6403-universal-darwin-17", "sorbet-static-0.5.6403-universal-darwin-16", "sorbet-static-0.5.6403-universal-darwin-15", "sorbet-static-0.5.6403-universal-darwin-14"]
HTTP GET https://index.rubygems.org/versions
HTTP 304 Not Modified https://index.rubygems.org/versions
Fetching gem metadata from https://rubygems.org/
Double checking for ["libv8", "sorbet-static", "sorbet"] in rubygems repository https://rubygems.org/ or installed locally
Fetching gem metadata from https://rubygems.org/
Looking up gems ["libv8", "sorbet-static", "sorbet"]
Could not find sorbet-static-0.5.6403-universal-darwin-20 in any of the sources
Bundler::GemNotFound: Could not find sorbet-static-0.5.6403-universal-darwin-20 in any of the sources
/home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/spec_set.rb:95:in `block in materialize'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/spec_set.rb:90:in `map!'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/spec_set.rb:90:in `materialize'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/definition.rb:183:in `specs'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/installer.rb:254:in `ensure_specs_are_compatible!'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/installer.rb:84:in `block in run'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/process_lock.rb:12:in `block in lock'
  <internal:core> core/io.rb:699:in `open'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/process_lock.rb:9:in `lock'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/installer.rb:72:in `run'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/installer.rb:24:in `install'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/cli/install.rb:64:in `run'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/cli.rb:259:in `block in install'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/settings.rb:134:in `temporary'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/cli.rb:258:in `install'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/cli.rb:30:in `dispatch'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/cli.rb:24:in `start'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/exe/bundle:49:in `block in <top (required)>'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/lib/bundler/friendly_errors.rb:130:in `with_friendly_errors'
  /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems/gems/bundler-2.2.17/exe/bundle:37:in `<top (required)>'
  <internal:core> core/kernel.rb:400:in `load'
  <internal:core> core/kernel.rb:400:in `load'
  /home/eregon/.rubies/truffleruby-jvm/bin/bundle:23:in `<main>'
```
Because it tries to resolve a darwin-specific spec and that seems to fail (this is on Linux, so it kind of makes sense I guess).

On Bundler 2.1.4 (the version shipped with TruffleRuby), `bundle install` works:
```
Fetching gem metadata from https://rubygems.org/.
Using bundler 2.1.4
Fetching libv8 8.4.255.0 (x86_64-linux)
Installing libv8 8.4.255.0 (x86_64-linux)
Fetching sorbet-static 0.5.6403 (x86_64-linux)
Installing sorbet-static 0.5.6403 (x86_64-linux)
Fetching sorbet 0.5.6403
Installing sorbet 0.5.6403
Bundle complete! 2 Gemfile dependencies, 4 gems now installed.
```
and the lockfile is:
```
GEM
  remote: https://rubygems.org/
  specs:
    libv8 (8.4.255.0)
    sorbet (0.5.6403)
      sorbet-static (= 0.5.6403)
    sorbet-static (0.5.6403-universal-darwin-14)

PLATFORMS
  ruby

DEPENDENCIES
  libv8 (= 8.4.255.0)
  sorbet (= 0.5.6403)

BUNDLED WITH
   2.1.4
```
(weird to see universal-darwin here, I'mn on Linux, and strange that libv8 doesn't show anything platform specific)

So this seems some kind of regression between Bundler 2.1.4 and 2.2.17, so maybe an existing issue in 2.14 that now shows up as a real failure for `bundle install`.

Something a bit special about TruffleRuby here is that `Gem.platforms` [is only `[RUBY]`](https://github.com/oracle/truffleruby/blob/04016dc0c101a0289d0a74049b4525e218b63337/lib/truffle/rubygems/defaults/truffleruby.rb#L23), because precompiled C extensions are generally not compatible (different ABI).
However, some gems do not depend on the Ruby ABI and then it's fine to install the binary gem on TruffleRuby. This is the case for `libv8` and for `sorbet-static` (handled here: https://github.com/oracle/truffleruby/blob/72ce69ec079618a3c0513da968262a4484633834/lib/mri/rubygems/platform.rb#L44).

## Environment

```
Bundler       2.2.17
  Platforms   ruby
Ruby          2.7.3p0 (2021-05-07 revision 04016dc0c101a0289d0a74049b4525e218b63337) [x86_64-linux]
  Full Path   /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/bin/truffleruby
  Config Dir  /home/eregon/code/truffleruby-ws/truffleruby/etc
RubyGems      3.1.6
  Gem Home    /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems
  Gem Path    /home/eregon/.gem/truffleruby/2.7.3.1:/home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/lib/gems
  User Home   /home/eregon
  User Path   /home/eregon/.gem/truffleruby/2.7.3.1
  Bin Dir     /home/eregon/code/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_077B8863DC_JAVA11/graalvm-077b8863dc-java11-21.2.0-dev/languages/ruby/bin
Tools         
  Git         2.30.2
  RVM         not installed
  rbenv       not installed
  chruby      0.3.9
```

## Bundler Build Metadata

```
Built At          2021-05-05
Git SHA           69cbd6e10e
Released Version  true
```

### Gemfile

```ruby
source 'https://rubygems.org'

gem 'sorbet', '0.5.6403'
gem 'libv8', '8.4.255.0'
```

### Gemfile.lock

```
<No /home/eregon/code/truffleruby-ws/truffleruby/Gemfile.lock found>
```

@tomstuart first noticed this bug.

**Comment by eregon** [comment]
@deivid-rodriguez Indeed, I am sorry, I was too eager (I think partly because I was excited I got a build to reproduce pretty quickly). I also hesitated to reply as I didn't want to spam you with too many notifications.

Do you think it's something you could have a look this or next week? If not, could you maybe answer some of the questions in https://github.com/rubygems/rubygems/issues/4588#issuecomment-836986049 and then I'll try to dig deeper?

**Comment by deivid-rodriguez** [comment]
What bundler 2.2 does is try to find an exact platform specific match. If it finds it, it locks that. Otherwise, it locks all available variants with the same specificity (so that the generated lockfile works by default on those platforms). In truffleruby's case, the local platform is RUBY and `sorbet-static` has no RUBY variant, so bundler resorts to locking all variants.

To fix this issue, we'd need to make bundler go through your custom logic in `Gem::Platform.match_spec?`. I think a patch like this does the trick:

```diff
diff --git a/bundler/lib/bundler/spec_set.rb b/bundler/lib/bundler/spec_set.rb
index e1012b9284..04f75df61c 100644
--- a/bundler/lib/bundler/spec_set.rb
+++ b/bundler/lib/bundler/spec_set.rb
@@ -199,7 +199,8 @@ def tsort_each_node
     def spec_for_dependency(dep, match_current_platform)
       specs_for_platforms = lookup[dep.name]
       if match_current_platform
-        GemHelpers.select_best_platform_match(specs_for_platforms, Bundler.local_platform).select{|s| Gem::Platform.match_spec?(s) }
+        GemHelpers.select_best_platform_match(specs_for_platforms.select{|s| Gem::Platform.match_spec?(s) }, Bundler.local_platform)
       else
         GemHelpers.select_best_platform_match(specs_for_platforms, dep.__platform)
       end
```

**Comment by tomstuart** [comment]
This patch does indeed fix the issue I was seeing.

**Comment by eregon** [comment]
I also verified on the minimal repro above and it works, thank you!
Do you think that change would be fine to merge to Bundler, or we need some other solution?
Could you make a PR? :smiley: 

I'm unsure if `Bundler.local_platform` being `Gem::Platform::RUBY` is right, given that `Gem::Platform.local` is OS-specific (e.g., `x86_64-linux`).
Maybe asking the local platform should also pass a gem name? That seems a bit surprising/unexpected though.
If it works well in practice, there is probably no much point to change that.
It's kind of nice it works like `force_ruby_platform`, although maybe that should have an effect on TruffleRuby too. However, it wouldn't be useful for sorbet-static & libv8, so seems best as-is currently.

Regarding the diff, I noticed there is filtering in `select_best_platform_match` that uses `spec.match_platform(platform)` and that has its own logic in `Bundler::MatchPlatform`. It'd be nice if `Gem::Platform.match_spec?` was used there, but there might be some differences in behavior besides the one we want in this case?

**Comment by deivid-rodriguez** [comment]
Yes, I'll make a PR.

> Regarding the diff, I noticed there is filtering in select_best_platform_match that uses spec.match_platform(platform) and that has its own logic in Bundler::MatchPlatform. It'd be nice if Gem::Platform.match_spec? was used there, but there might be some differences in behavior besides the one we want in this case?

Yes, that's what I intuitively thought too but that logic works against custom platforms, not only against the local platform, and `Gem::Platform.match_spec?` is specific to the local platform. It feels that this could be done in a cleaner way, but I think we can add the quick patch to fix this issue and consider improving things in the future.

**Comment by deivid-rodriguez** [comment]
> I'm unsure if Bundler.local_platform being Gem::Platform::RUBY is right, given that Gem::Platform.local is OS-specific (e.g., x86_64-linux).

Normally `Bundler.local_platform` is OS-specific too unless `forcce_ruby_platform` is set. We only add some extra logic to make it `Gem::Platform::RUBY` if `Gem.platforms` only includes `Gem::Platform::RUBY` to fix another issue with `truffleruby`.

**Comment by eregon** [comment]
> Yes, that's what I intuitively thought too but that logic works against custom platforms, not only against the local platform, and `Gem::Platform.match_spec?` is specific to the local platform.

Indeed, I missed that.
Maybe a good API then would be something like `Bundler.local_platform_for(gem)` / `Gem::Platform.local_platform_for(gem)`, and replace `Bundler.local_platform` by that in places where it makes sense?

> We only add some extra logic to make it `Gem::Platform::RUBY` if `Gem.platforms` only includes `Gem::Platform::RUBY` to fix another issue with `truffleruby`.

Right, that's since https://github.com/rubygems/rubygems/pull/4333 and it seems safer in general.

**Comment by deivid-rodriguez** [comment]
>  Maybe a good API then would be something like Bundler.local_platform_for(gem) / Gem::Platform.local_platform_for(gem), and replace Bundler.local_platform by that in places where it makes sense?

Yeah, I was thinking the same thing :+1:.

**Comment by deivid-rodriguez** [comment]
It was actually released with 2.2.23, where a proper entry was included in the [release notes](https://github.com/rubygems/rubygems/releases/tag/bundler-v2.2.23).

**Comment by eregon** [comment]
Oh I missed that, thanks!
